src/RiakClient/Commands/CommandBuilder{TBuilder,TCommand,TOptions}.cs

```diff
@@ -0,0 +1,115 @@
+﻿// <copyright file="CommandBuilder{TBuilder,TCommand,TOptions}.cs" company="Basho Technologies, Inc.">
+// Copyright 2015 - Basho Technologies, Inc.
+//
+// This file is provided to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file
+// except in compliance with the License.  You may obtain
+// a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+// </copyright>
+
+namespace RiakClient.Commands
+{
+    using System;
+
+    /// <summary>
+    /// Base class for all Riak command builders.
+    /// </summary>
+    /// <typeparam name="TBuilder">The type of the builder. Allows chaining.</typeparam>
+    /// <typeparam name="TCommand">The type of the command.</typeparam>
+    /// <typeparam name="TOptions">The type of the options for this command.</typeparam>
+    public abstract class CommandBuilder<TBuilder, TCommand, TOptions>
+        where TBuilder : CommandBuilder<TBuilder, TCommand, TOptions>
```

Comment:
This allows chaining by passing the most-derived type as a generic. Each `WithFoo()` method returns the current object cast to the most-derived type:

```csharp
public TBuilder WithFoo(string yomama)
{
    this.yomama = yomama;
    return (TBuilder)this;
}
```

Without this, you can't have `With...()` methods that chain in any order.